src/Native/CpuMathNative/Sse.cpp

```diff
@@ -60,17 +60,166 @@ const unsigned int TrailingAlignmentMask[16] =
+
+                        const float* pMatTemp = pMatCurrent;
+                        __m128 x01 = _mm_mul_ps(vector, _mm_load_ps(pMatTemp));
+                        __m128 x11 = _mm_mul_ps(vector, _mm_load_ps(pMatTemp += ccol));
+                        __m128 x21 = _mm_mul_ps(vector, _mm_load_ps(pMatTemp += ccol));
+                        __m128 x31 = _mm_mul_ps(vector, _mm_load_ps(pMatTemp += ccol));
+
+                        res0 = _mm_add_ps(res0, x01);
+                        res1 = _mm_add_ps(res1, x11);
+                        res2 = _mm_add_ps(res2, x21);
+                        res3 = _mm_add_ps(res3, x31);
+
+                        pSrcCurrent += 4;
+                        pMatCurrent += 4;
+                    }
+                }
+                else
+                {
+                    // Handle the "worst-case" scenario, which is when we have 4-8 elements and the input is not
+                    // 128-bit aligned. This means we can't do any aligned loads and will just end up doing two
+                    // unaligned loads where we mask the input each time.
+                    remainder = length;
+                }
+
+                if (remainder != 0)
+                {
+                    // Handle any trailing elements that don't fit into a 128-bit block by moving back so that the next
+                    // unaligned load will read to the end of the array and then mask out any elements already processed
+
+                    pMatCurrent -= (4 - remainder);
+                    pSrcCurrent -= (4 - remainder);
+
+                    __m128 mask = _mm_loadu_ps(((float*)(&TrailingAlignmentMask)) + (remainder * 4));
+
+                    const float* pMatTemp = pMatCurrent;
+                    __m128 x01 = _mm_and_ps(mask, _mm_loadu_ps(pMatTemp));
+                    __m128 x11 = _mm_and_ps(mask, _mm_loadu_ps(pMatTemp += ccol));
+                    __m128 x21 = _mm_and_ps(mask, _mm_loadu_ps(pMatTemp += ccol));
+                    __m128 x31 = _mm_and_ps(mask, _mm_loadu_ps(pMatTemp += ccol));
+                    __m128 vector = _mm_and_ps(mask, _mm_loadu_ps(pSrcCurrent));
+
+                    res0 = _mm_add_ps(res0, _mm_mul_ps(x01, vector));
+                    res1 = _mm_add_ps(res1, _mm_mul_ps(x11, vector));
+                    res2 = _mm_add_ps(res2, _mm_mul_ps(x21, vector));
+                    res3 = _mm_add_ps(res3, _mm_mul_ps(x31, vector));
+
+                    pMatCurrent += 4;
+                    pSrcCurrent += 4;
+                }
+            }
+
+            // Add up the entries of each, with the 4 results in res0
+            res0 = _mm_hadd_ps(res0, res1);
+            res2 = _mm_hadd_ps(res2, res3);
+            res0 = _mm_hadd_ps(res0, res2);
+
+            _mm_storeu_ps(pDstCurrent, res0);
+
+            pDstCurrent += 4;
+            pMatCurrent += 3 * ccol;
+        }
+    }
+    
     while (pDstCurrent < pDstEnd)
```

Comment:
Why are we only doing one at a time, rather than doing 4 at a time until we can't anymore and then falling back to 1 at a time for the (at most) 3 trailing elements?

Reply:
> rather than doing 4 at a time until we can't anymore and then falling back to 1 at a time for the (at most) 3 trailing elements

we are following this approach when ccol is a multiple of 4 because then if pmat is aligned, will mean that ````pmat  +  k * ccol``` will be aligned and we can use the loadAligned for all four rows which wont be the case if ccol is not a multiple of 4, hence we will have calculate the misalignment for each row independently

Reply:
We could just move the ccol check as part of the `if ((misalignment & 3) != 0)` check.

Operating on 4 columns at a time, aligned or not, will likely perform slightly better and guarantees that we have, at most, 3 trailing elements which makes the trailing code simpler as well (you can just manually unroll it)

Reply:
is the unrolling okay ?

```C#
 {
     switch (numRows)
    {
        case 3:
            *(pDstCurrent + 2) =  RowMul(pMatCurrent + 2 * ccol , psrc, pSrcEnd);
        case 2:
            *(pDstCurrent + 1) =  RowMul(pMatCurrent + ccol, psrc, pSrcEnd);
        case 1:
            *pDstCurrent =  RowMul(pMatCurrent, psrc, pSrcEnd);
    }
}
float RowMul(_In_ const float* pMatCurrent, _In_ const float* pSrcCurrent, _In_ const float* pSrcEnd)
{
    __m128 res0 = _mm_setzero_ps();
    while (pSrcCurrent < pSrcEnd)
    {
        __m128 vector = _mm_loadu_ps(pSrcCurrent);

        const float* pMatTemp = pMatCurrent;
        __m128 x01 = _mm_mul_ps(vector, _mm_loadu_ps(pMatTemp));
        res0 = _mm_add_ps(res0, x01);

        pSrcCurrent += 4;
        pMatCurrent += 4;
    }
    
    // Sum all the elements together and return the result
    res0 = _mm_hadd_ps(res0, res0);
    res0 = _mm_hadd_ps(res0, res0);
    return _mm_cvtss_f32(res0);
}

```

Reply:
Wasn't there a code snippet here? I was writing up a response to it...

Reply:
sry I edited the snippet. earlier was failing the tests
```C#
    {
                switch (numRows)
                {
                    case 3:
                        *(pDstCurrent + 2) = RowMultiply(pMatCurrent + 2 * ccol, psrc, pSrcEnd, ccol);
                        goto case 2;
                    case 2:
                        *(pDstCurrent + 1) = RowMultiply(pMatCurrent + 1 * ccol, psrc, pSrcEnd, ccol);
                        goto case 1;
                    case 1:
                        *pDstCurrent = RowMultiply(pMatCurrent, psrc, pSrcEnd, ccol);
                        break;
                }
            }
        }

        private  static unsafe float RowMultiply(float* pMatCurrent, float* pSrcCurrent, float* pSrcEnd, int ccol)
        {
            Vector256<float> res0 = Avx.SetZeroVector256<float>();
            int remainder = ccol % 8;
            while (pSrcCurrent + 8 <= pSrcEnd)
            {
                Vector256<float> vector = Avx.LoadVector256(pSrcCurrent);

                float* pMatTemp = pMatCurrent;
                res0 = MultiplyAdd(pMatTemp, vector, res0);

                pSrcCurrent += 8;
                pMatCurrent += 8;
            }

            res0 = VectorSum256(in res0);
            float sum = Sse.ConvertToSingle(Sse.AddScalar(Avx.GetLowerHalf(res0), GetHigh(res0)));

            switch (remainder)
            {
                case 7: sum += *(pSrcCurrent + 6) * *(pMatCurrent + 6); goto case 6;
                case 6: sum += *(pSrcCurrent + 5) * *(pMatCurrent + 5); goto case 5;
                case 5: sum += *(pSrcCurrent + 4) * *(pMatCurrent + 4); goto case 4;
                case 4: sum += *(pSrcCurrent + 3) * *(pMatCurrent + 3); goto case 3;
                case 3: sum += *(pSrcCurrent + 2) * *(pMatCurrent + 2); goto case 2;
                case 2: sum += *(pSrcCurrent + 1) * *(pMatCurrent + 1); goto case 1;
                case 1: sum += *(pSrcCurrent) * *(pMatCurrent); break;
            }

            return sum;
        }
```